File: AlgorithmPractice/src/Programmers/kakaoBLIND2021/메뉴리뉴얼/Solution.java
package Programmers.kakaoBLIND2021.메뉴리뉴얼;

import java.util.*;
class Solution {
    
    Map<String,Integer> map  = new TreeMap<>();
    public String[] solution(String[] orders, int[] course) {
        String[] answer = {};

        for(int i=0;i<orders.length;i++){
            char tmp[] = orders[i].toCharArray();
            Arrays.sort(tmp);
            orders[i]=new String(tmp);
            bit(tmp,tmp.length,i);
        }
        
        Set<String> tmp_set = new HashSet<>(map.keySet());
        
        for(String str : tmp_set){
            if(map.get(str)<2)
                map.remove(str);
        }
        
        Set<String> ret = new TreeSet<>();
        
        for(int cnt:course){
            
            int max=0;
            
            for(String str : map.keySet()){
                if(str.length()==cnt){
                    max=Math.max(max,map.get(str));
                }
            }
            
            for(String str : map.keySet()){
                if(str.length()==cnt&&map.get(str)==max)
                    ret.add(str);
            }
        }
        
        
        return ret.toArray(new String[0]);
    }
    
    public String[] bit(char[] arr, int n, int idx) {
    List<String> list = new ArrayList<>();
  
    for(int i=0; i < 1<<n; i++) {
        StringBuilder sb  =new StringBuilder();
        for(int j=0; j<n; j++) {
            if((i & 1<<j) != 0) 
                sb.append(arr[j]);
        }
        if(sb.length()>=2){
            String tmp = sb.toString();
            map.put(tmp,map.getOrDefault(tmp,0)+1);
        }
        
        
    }
        return list.toArray(new String[0]);
    }
    
}

File: AlgorithmPractice/bin/Programmers/kakaoBLIND2021/메뉴리뉴얼/solution.py
from itertools import combinations as combi

def solution(orders, course):
    answer = []
    d1 = {}
    d2 = {}
    for order in orders:
        l = len(order)
        order=sorted(order)
        for i in range(2,l+1):
            for j in combi(order,i):
                
                key = ''.join(j)
                d1[key] = d1.get(key,0)+1
    
    for k,v in d1.items():
        if len(k) in course and v>=2:
            t = d2.get(len(k),list())
            t.append((k,v))
            d2[len(k)]=t
            
    for k,v in d2.items():
        v.sort(key=lambda x : -x[1])
        _max = v[0][1]
        
        for i in v:
            if i[1]==_max:
                answer.append(i[0])
            else:
                break
    return sorted(answer)